pkg/certificates/bootstrap/reader.go

```diff
@@ -0,0 +1,392 @@
+			Factor:   2,
+			Jitter:   0.1,
+			Steps:    5,
+		}
+		if err := wait.ExponentialBackoff(backoff, s.rotateCerts); err != nil {
+			utilruntime.HandleError(fmt.Errorf("Reached backoff limit, still unable to rotate certs: %v", err))
+			wait.PollInfinite(32*time.Second, s.rotateCerts)
+		}
+
+	}, time.Second, s.stopCh)
+}
+
+func (s *SelfSignManager) Stop() {
+	s.certAccessLock.Lock()
+	defer s.certAccessLock.Unlock()
+	if s.stopped {
+		return
+	}
+	close(s.stopCh)
+	s.stopped = true
+}
+
+func (s *SelfSignManager) Current() *tls.Certificate {
+	s.certAccessLock.Lock()
+	defer s.certAccessLock.Unlock()
+	return s.cert
+}
+
+func (s *SelfSignManager) CA() *x509.Certificate {
+	s.certAccessLock.Lock()
+	defer s.certAccessLock.Unlock()
+	return s.caCert
+}
+
+func (s *SelfSignManager) CABundle() *x509.CertPool {
+	s.certAccessLock.Lock()
+	defer s.certAccessLock.Unlock()
+	return s.caBundle
+}
+
+func (s *SelfSignManager) ServerHealthy() bool {
+	s.certAccessLock.Lock()
+	defer s.certAccessLock.Unlock()
+	return s.serverHealth
+}
+
+func NewSelfSignedCertManager(config *SelfSignManagerConfig) (CertificateManager, error) {
+	return &SelfSignManager{config: config}, nil
+}
+
+func (s *SelfSignManager) validateSecret(secret *k8sv1.Secret) error {
+	if _, ok := secret.Data[CertBytesValue]; !ok {
+		return fmt.Errorf("%s value not found in %s secret\n", CertBytesValue, s.config.SecretName)
+	}
+	if _, ok := secret.Data[KeyBytesValue]; !ok {
+		return fmt.Errorf("%s value not found in %s secret\n", KeyBytesValue, s.config.SecretName)
+	}
+	if _, ok := secret.Data[SigningCertBytesValue]; !ok {
+		return fmt.Errorf("%s value not found in %s secret\n", SigningCertBytesValue, s.config.SecretName)
+	}
+	return nil
+}
+
+func (s *SelfSignManager) loadCertificates(secret *k8sv1.Secret) (serverCrt *tls.Certificate, caCrt *x509.Certificate, err error) {
+
+	if err := s.validateSecret(secret); err != nil {
+		return nil, nil, err
+	}
+
+	crt, err := tls.X509KeyPair(secret.Data[CertBytesValue], secret.Data[KeyBytesValue])
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load certificate: %v\n", err)
+	}
+	leaf, err := cert.ParseCertsPEM(secret.Data[CertBytesValue])
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load leaf certificate: %v\n", err)
+	}
+	crt.Leaf = leaf[0]
+	cas, err := cert.ParseCertsPEM(secret.Data[SigningCertBytesValue])
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load ca certificate: %v\n", err)
+	}
+	return &crt, cas[0], nil
+}
+
+func (s *SelfSignManager) rotateCerts() (bool, error) {
+
+	obj, exists, _ := s.informer.GetStore().GetByKey(s.config.Namespace + "/" + s.config.SecretName)
+	if !exists {
+		return false, nil
+	}
+	secret := obj.(*k8sv1.Secret)
+
+	if s.resourceVersion == secret.ResourceVersion {
+		return true, nil
+	}
+
+	crt, ca, err := s.loadCertificates(secret)
+	if err != nil {
+		fmt.Println(err)
+		return false, nil
+	}
+
+	if s.config.AfterCertRotateCallback != nil {
```

Comment:
what is this for? it doesn't look like anything implements this callback.

Reply:
Just a leftover.